zingo-sync/src/sync.rs

```diff
@@ -27,62 +31,107 @@ pub async fn sync<P, W>(
     let fetcher_handle = tokio::spawn(fetch(fetch_request_receiver, client));
+    handles.push(fetcher_handle);
+
+    update_scan_ranges(
+        fetch_request_sender.clone(),
+        parameters,
+        wallet.get_birthday().unwrap(),
+        wallet.get_sync_state_mut().unwrap(),
+    )
+    .await
+    .unwrap();
+
+    // create channel for receiving scan results and launch scanner
+    let (scan_results_sender, mut scan_results_receiver) = mpsc::unbounded_channel();
+    let ufvks = wallet.get_unified_full_viewing_keys().unwrap();
+    let mut scanner = Scanner::new(
+        scan_results_sender,
+        fetch_request_sender,
+        parameters.clone(),
+        ufvks,
+    );
+    scanner.spawn_workers();
+
+    let mut interval = tokio::time::interval(Duration::from_millis(30));
+    loop {
+        interval.tick().await;
+
+        // if a scan worker is idle, send it a new scan task
+        if scanner.is_worker_idle() {
+            if let Some(scan_range) = prepare_next_scan_range(wallet.get_sync_state_mut().unwrap())
+            {
+                let previous_wallet_block = wallet
+                    .get_wallet_block(scan_range.block_range().start - 1)
+                    .ok();
+
+                scanner
+                    .add_scan_task(ScanTask::from_parts(scan_range, previous_wallet_block))
+                    .unwrap();
+            } else {
+                // when no more ranges are available to scan, break out of the loop
+                break;
+            }
+        }
+
+        match scan_results_receiver.try_recv() {
+            Ok((scan_range, scan_results)) => {
+                update_wallet_data(wallet, scan_results).unwrap();
+                // TODO: link nullifiers and scan linked transactions
+                remove_irrelevant_data(wallet, &scan_range).unwrap();
+                mark_scanned(scan_range, wallet.get_sync_state_mut().unwrap()).unwrap();
+            }
+            Err(TryRecvError::Empty) => (),
+            Err(TryRecvError::Disconnected) => break,
+        }
+    }
 
-    update_scan_ranges(fetch_request_sender.clone(), parameters, &sync_state)
-        .await
-        .unwrap();
-
-    let account_ufvks = wallet.get_unified_full_viewing_keys().unwrap();
-    let scanning_keys = ScanningKeys::from_account_ufvks(account_ufvks);
-
-    let scan_range = prepare_next_scan_range(&sync_state);
-    if let Some(range) = scan_range {
-        scan(
-            fetch_request_sender,
-            parameters,
-            &scanning_keys,
-            range.clone(),
-        )
-        .await
-        .unwrap();
+    drop(scanner);
+    while let Some((scan_range, scan_results)) = scan_results_receiver.recv().await {
+        update_wallet_data(wallet, scan_results).unwrap();
+        // TODO: link nullifiers and scan linked transactions
+        remove_irrelevant_data(wallet, &scan_range).unwrap();
+        mark_scanned(scan_range, wallet.get_sync_state_mut().unwrap()).unwrap();
     }
```

Comment:
DRY this block?

Reply:
yes if we keep this repeated part for shutdown we can DRY, i havent yet because i dont know if this architecture will stay and I also dont know what this block really is yet